File: 0001~0500/0070. Climbing Stairs/main.cpp
// Source: https://leetcode.com/problems/climbing-stairs/
// Date: 2021/7/11
// Constraints: 1 <= n <= 45

#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int climbStairs(int n) {
        vector<int>staircase(n+1);
        staircase[0] = 1;
        staircase[1] = 1;
        for(int i=2; i<=n; i++) 
            staircase[i] = staircase[i-1]+staircase[i-2];
        return staircase[n];
    }
};

File: 0001~0500/0070. Climbing Stairs/main.py
"""
Source: https://leetcode.com/problems/climbing-stairs/description/
Date: 2022/12/12
Skill: 
Runtime: 54 ms, faster than 51.57% 
Memory Usage: 13.7 MB, less than 96.52%
Time complexity: 
Space complexity: 
Constraints: 
    1 <= n <= 45
"""

from typing import List, Optional
import collections

class Solution:
    def climbStairs(self, n: int) -> int:
        prev, cur = 1, 1
        for i in range(2, n+1):
            prev, cur = cur, prev+cur
        return cur
        